vmagent: support MDX(monitor data exchange) for remote write (#10854)
VictoriaMetrics/VictoriaMetrics: docs/victoriametrics/vmagent.md

  ```sh
  ./vmagent -remoteWrite.url=http://remote-storage/api/v1/write -streamAggr.dropInputLabels=replica -streamAggr.dedupInterval=60s
  ```
  
### Monitoring Data eXchange

> The MDX (Monitoring Data eXchange) is a monitoring of monitoring metrics collection and sharing feature. 

Reviewer: This sentence explains one unfamiliar concept (MDX) by referring to another unfamiliar concept, "monitoring of monitoring". A reader is unlikely to know either, so it doesn't really improve understanding.

MDX should be explained using terms that are already familiar to the average user. It may make sense to merge this sentence with the next one and introduce the concept in a more direct and concrete way.

It aims to send only metrics from the VictoriaMetrics services to the corresponding `-remoteWrite.url`, discarding metrics from non-VictoriaMetrics services.

When enabling MDX for the `-remoteWrite.url`, `vmagent` will only forward metrics from the instances that emit `vm_app_version`, which is a metric that all VictoriaMetrics services will emit.
The number of dropped rows from non-VictoriaMetrics services is exposed as `vmagent_remotewrite_mdx_rows_dropped_total`.

To enable MDX, set `-remoteWrite.mdx.enable=true` for the target URL and `-remoteWrite.mdx.enable=false` for other URLs:

```sh
./vmagent \
  -remoteWrite.url=http://service-to-keep-all-metrics:8428/api/v1/write \
  -remoteWrite.mdx.enable=false \
  -remoteWrite.url=http://service-to-keep-only-vm-metrics:8428/api/v1/write \
  -remoteWrite.mdx.enable=true 
```

> `vmagent` relies on the `vm_app_version` metric to identify the VictoriaMetrics instance. `vmagent` will not recognize an instance as a VictoriaMetrics instance and retain the instance's metrics until it receives the `vm_app_version` metric from this instance.

Reviewer: Sorry, I cannot get the meaning here.

  ```sh
  ./vmagent -remoteWrite.url=http://remote-storage/api/v1/write -streamAggr.dropInputLabels=replica -streamAggr.dedupInterval=60s
  ```
  
### Monitoring Data eXchange

> The MDX (Monitoring Data eXchange) feature aims to send only metrics from the VictoriaMetrics services to the corresponding `-remoteWrite.url`, discarding metrics from non-VictoriaMetrics services.

Reviewer: ```suggestion
The Monitoring Data eXchange (MDX) feature allows vmagent to forward only VictoriaMetrics metrics to selected `-remoteWrite.url` destinations while dropping metrics from non-VictoriaMetrics services.
```

To enable MDX, set `-remoteWrite.mdx.enable=true` for the target URL and `-remoteWrite.mdx.enable=false` for other URLs:

```sh
./vmagent \
  -remoteWrite.url=http://service-to-keep-all-metrics:8428/api/v1/write \
  -remoteWrite.mdx.enable=false \
  -remoteWrite.url=http://service-to-keep-only-vm-metrics:8428/api/v1/write \
  -remoteWrite.mdx.enable=true 
```

> When enabling MDX for the `-remoteWrite.url`, `vmagent` will only forward the metrics from the instances that emit `vm_app_version`, which is a metric that all VictoriaMetrics services will emit,

Reviewer: what about:

```
When MDX is enabled for a `-remoteWrite.url`, vmagent forwards only metrics from targets that:
- expose the `vm_app_version` metric (emitted by all VictoriaMetrics components)
- contain the label specified via `-mdx.label`.
```

later we can add `victoriametrics_app=true` to the list

or the metrics contain the label specified by `mdx.label`:

```sh
./vmagent \
  -remoteWrite.url=http://service-to-keep-only-vm-metrics:8428/api/v1/write \
  -remoteWrite.mdx.enable=true \
  -mdx.label="service=victoriametrics"
```

> 

Reviewer: ```suggestion
In this configuration, metrics with the label `service=victoriametrics` are preserved even if their scrape targets do not expose `vm_app_version` metric.
```

> The number of preserved rows from non-VictoriaMetrics services is exposed as `vmagent_remotewrite_mdx_rows_preserved_total`.

Reviewer: ```suggestion
The number of VictoriaMetrics metrics preserved by MDX is exposed as `vmagent_remotewrite_mdx_rows_preserved_total`.
```

      H1 --> H2[per-url <a href="https://docs.victoriametrics.com/victoriametrics/relabeling/">relabeling</a><br><b>-remoteWrite.urlRelabelConfig</b>]
      H2 --> H3[per-url <a href="https://docs.victoriametrics.com/victoriametrics/stream-aggregation">aggregation</a><br><b>-remoteWrite.streamAggr.config</b><br><b>-remoteWrite.streamAggr.dedupInterval</b>]
      H3 --> H4["per-url <a href="https://docs.victoriametrics.com/victoriametrics/vmagent/#calculating-disk-space-for-persistence-queue">queue</a> (default: enabled)<br><b>-remoteWrite.disableOnDiskQueue</b>"]
      H4 --> H5[<a href="https://docs.victoriametrics.com/victoriametrics/vmagent/#adding-labels-to-metrics">add extra labels</a><br><b>-remoteWrite.label</b>]
      H5 --> H6[[push to <b>-remoteWrite.url</b>]]

      %% Right branch
      G --> R1[per-url <a href="https://docs.victoriametrics.com/victoriametrics/vmagent/#monitoring-data-exchange">mdx filter</a><br><b>-remoteWrite.mdx.enable</b>]
      R1 --> R2[per-url <a href="https://docs.victoriametrics.com/victoriametrics/relabeling/">relabeling</a><br><b>-remoteWrite.urlRelabelConfig</b>]

>       R2 --> R3[per-url <a href="https://docs.victoriametrics.com/victoriametrics/stream-aggregation">aggregation</a><br><b>-remoteWrite.streamAggr.config</b><br><b>-remoteWrite.streamAggr.dedupInterval</b>]

Reviewer: IMO, the persistent queue should not be present on the Life of the Sample diagram because it has little to do with the sample. 

@hagen1778 wdyt

<img width="647" height="688" alt="Image" src="https://github.com/user-attachments/assets/4548fa9f-fd7b-47bb-8eff-154ac5391981" />

Reviewer 2: I am okay with having -remoteWrite.disableOnDiskQueue on this diagram. It has the same relation to life of a sample as `-maxIngestionRate `, complexity or cardinality limiting. 

I am not sure if MDX filter should be there, because its use case is rather limited for a common user. But I am ok with that too, it needs to be documented anyway. 
What concerns me is that MDX goes after sharding, relabeling, complexity and cardinality limiting, stream aggregation. Isn't this affecting reliability of this setting? If nothing can be done about this, it should be at least highlighted in MDX docs.

Author: Yeah MDX works at per-url level, so it's hard to have it works before global mechanisms. I documented its working stage
